nebula-logger/core/main/logger-engine/classes/LoggerStackTrace.cls

```diff
@@ -0,0 +1,404 @@
+            }
+        }
+    }
+
+    private AnonymousBlock generateAnonymousBlock(String cleanedLocation) {
+        AnonymousBlock anonymousBlock = new AnonymousBlock();
+        return anonymousBlock;
+    }
+
+    private static String getNamespacePrefix() {
+        String className = LoggerStackTrace.class.getName();
+        return className.contains('.') ? className.substringBefore('.') : '';
+    }
+
+    private ApexClass generateApexClass(String cleanedLocation) {
+        ApexClass apexClass = new ApexClass();
+
+        List<String> locationPieces = cleanedLocation.split('\\.');
+
+        // It's nearly impossible to distinguish between a namespaced class vs an inner-class
+        // Example: Nebula.Logger.saveLog() vs SomeClass.SomeInnerClass.someMethod().
+        // This can cause some issues for Nebula Logger's managed package, so compare
+        // to the namespace to determine if it's a namespaced class.
+        String apexTopLevelClassName;
+        if (locationPieces.get(0) == getNamespacePrefix()) {
+            apexTopLevelClassName = locationPieces.get(0) + '.' + locationPieces.get(1);
+        } else {
+            apexTopLevelClassName = locationPieces.get(0);
+        }
+
+        String apexMethodName = locationPieces.size() >= 2 ? locationPieces.get(locationPieces.size() - 1) : null;
+        String apexInnerClassName;
+        if (String.isBlank(apexMethodName)) {
+            apexInnerClassName = cleanedLocation.substringAfter(apexTopLevelClassName + '.');
+        } else {
+            apexInnerClassName = cleanedLocation.substringAfter(apexTopLevelClassName + '.').substringBefore('.' + apexMethodName);
+            apexInnerClassName = apexInnerClassName == apexMethodName ? null : apexInnerClassName;
+        }
+
+        // For inner class methods - 'like MyClass.SomeInnerClass.myMethod' - the top-level class is stored
+        // as the API name, and the inner class name + method name (e.g., 'SomeInnerClass.myMethod') are stored
+        // as the action name
+        String qualifiedMethodName = '';
+        if (apexInnerClassName != null) {
+            qualifiedMethodName = apexInnerClassName + '.';
+        }
+        qualifiedMethodName += apexMethodName;
+
+        apexClass.ApiName = apexTopLevelClassName;
+        apexClass.ActionName = qualifiedMethodName;
+
+        return apexClass;
+    }
+
+    private ApexTrigger generateApexTrigger(String cleanedLocation) {
+        ApexTrigger apexTrigger = new ApexTrigger();
+        apexTrigger.ApiName = cleanedLocation;
+        return apexTrigger;
+    }
+
+    private static Boolean isValidStackTrace(String stackTraceString) {
+        // Some lines, especially in the managed package, can result in '()' or '(Some_Namespace)'
+        if (
+            String.isBlank(stackTraceString) ||
+            stackTraceString == '()' ||
+            System.Pattern.compile('^\\([0-9A-Za-z_]+\\)$').matcher(stackTraceString).matches()
```

Comment:
if this method is going to be called multiple times, lazy-loading the `Pattern` just once could be good

Reply:
Yeah, it'll definitely get called multiple times - I'll switch to a lazy-loaded constant for it